src/Microsoft.AspNetCore.Razor.Evolution/DefaultRazorSourceDocument.cs

```diff
@@ -23,17 +24,47 @@ public DefaultRazorSourceDocument(MemoryStream stream, Encoding encoding, string
             Filename = filename;
         }
 
-        public Encoding Encoding { get; }
+        private string Content
+        {
+            get
+            {
+                if (_content == null)
+                {
+                    InitializeContent();
+                }
+
+                return _content;
+            }
+        }
+
+        public override char this[int position] => Content[position];
+
+        public override Encoding Encoding { get; }
 
         public override string Filename { get; }
 
-        public override TextReader CreateReader()
+        public override int Length => Content.Length;
+
+        public override void CopyTo(int sourceIndex, char[] destination, int destinationIndex, int count) =>
+            Content.CopyTo(sourceIndex, destination, destinationIndex, count);
+
+        private void InitializeContent()
         {
             var copy = new MemoryStream(_stream.ToArray());
```

Comment:
`new StreamReader(copy, detectEncodingFromByteOrderMarks: true)`.

If I expand it to take in the current `Encoding` it explodes when `Encoding` is `null`. All this doesn't matter if you're [suggestion below](https://github.com/aspnet/Razor/pull/861#discussion_r88122755) works.